clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -253,15 +261,77 @@ void run(long now) {
         long pollTimeout = Math.min(result.nextReadyCheckDelayMs, notReadyTimeout);
         if (!result.readyNodes.isEmpty()) {
             log.trace("Nodes with data ready to send: {}", result.readyNodes);
+            // if some partitions are already ready to be sent, the select time would be 0;
+            // otherwise if some partition already has some data accumulated but not ready yet,
+            // the select time will be the time difference between now and its linger expiry time;
+            // otherwise the select time will be the time difference between now and the metadata expiry time;
             pollTimeout = 0;
         }
         sendProduceRequests(batches, now);
 
-        // if some partitions are already ready to be sent, the select time would be 0;
-        // otherwise if some partition already has some data accumulated but not ready yet,
-        // the select time will be the time difference between now and its linger expiry time;
-        // otherwise the select time will be the time difference between now and the metadata expiry time;
-        this.client.poll(pollTimeout, now);
+        return pollTimeout;
+
+    }
+
+    private boolean maybeSendTransactionalRequest(long now) {
+        if (transactionState != null && transactionState.hasInflightTransactionalRequest())
+            return true;
+
+        if (transactionState == null || !transactionState.hasPendingTransactionalRequests())
+            return false;
+
+        TransactionState.TransactionalRequest nextRequest = transactionState.nextTransactionalRequest();
+
+        if (nextRequest.isEndTxnRequest() && transactionState.isCompletingTransaction() && accumulator.hasUnflushedBatches()) {
+            if (!accumulator.flushInProgress())
+                accumulator.beginFlush();
+            transactionState.didNotSend(nextRequest);
+            return false;
+        }
+
+        Node targetNode = null;
+        long expiryTime = now + requestTimeout;
```

Comment:
Good point. This code will be invoked repeatedly until the transactional request succeeds. So no need to have any expiry time.